### PR TITLE
fix!: eliminate unnecessary merkle proofs in ADVZ shares

### PR DESCRIPTION
**Pull request by ggutoski**
closes: #671 

### This PR: 
- `Advz::evals_proof` is no longer a `Vec`
- structure of the merkle tree is changed accordingly. Each leaf in the merkle tree is a vec of polynomial evaluations as before, except now that vec looks like: https://github.com/EspressoSystems/jellyfish/blob/689555e07860c5ed6e54b467c76d9a1776f0f7b5/vid/src/advz.rs#L702-L715
- delete test `sad_path_verify_share_with_multiplicity` as it is a regression for #654 , which is no longer relevant as there is now only one merkle path in each share. 

### Key places to review:

 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added relevant changelog entries to the `CHANGELOG.md` of touched crates.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: vid/src/advz.rs [hunk 1]
```diff
@@ -515,35 +496,25 @@ where
...
-                    .iter()
-                    .map(FieldMultiplier),
-                pseudorandom_scalar,
-            );
+        let verification_iter = parallelizable_slice_iter(&multiplicities).map(|m| {
+            let evals_iter = evals.iter().skip(*m).step_by(multiplicity_usize);
+            let aggregate_eval =
+                polynomial_eval(evals_iter.map(FieldMultiplier), pseudorandom_scalar);
+            let domain_index = usize::try_from(share.index).map_err(vid)?
+                + m * usize::try_from(self.num_storage_nodes).map_err(vid)?;
```
Comment (akonring) [inline vid/src/advz.rs:511]: nit: we could use "convenience quantities" here as well.

File: vid/src/advz/test.rs [hunk 2]
```diff
@@ -254,51 +263,9 @@ fn verify_share_with_multiplicity() {
...
-        num_storage_nodes: 20,
-        max_multiplicity: 32, // payload fitting into a single polynomial
-        payload_len: 8200,
-    };
-    let (mut advz, payload) = advz_init_with::<Bn254>(advz_params);
-
-    let disperse = advz.disperse(payload).unwrap();
-    let (shares, common, commit) = (disperse.shares, disperse.common, disperse.commit);
-    for (i, share) in shares.iter().enumerate() {
-        // corrupt the last evaluation of the share
```
Comment (akonring) [inline vid/src/advz/test.rs:275]: Do we maybe want to keep some version of this test? Seems important to exercise the sad path for `multiplicity > 1`

File: vid/src/advz.rs [hunk 3]
```diff
@@ -715,27 +689,41 @@ where
...
 
-            // distribute evals from each poly among the storage nodes
+            // Populate a Vec of polynomial evaluations for all storage nodes:
+            //
+            // The `i`th item is a Vec of `polys.len() * multiplicity`
+            // polynomial evaluations.
             //
-            // perf warning: runtime is quadratic in payload_size
-            let mut all_storage_node_evals = vec![Vec::with_capacity(polys.len()); code_word_size];
+            // Evaluations for storage node `i` are ordered as follows. Define
```
Comment (akonring) [inline vid/src/advz.rs:706]: Great doc!

**Review by akonring — APPROVED** [review]
Nice.
Don't fully understand the motivation for moving to a different "share evaluation layout".
But they seem equivalent and the docs added are great and needed. Thanks.

File: vid/src/advz.rs (hunk 3 above)
Comment (ggutoski) [reply vid/src/advz.rs:706]: Thanks!

> Don't fully understand the motivation for moving to a different "share evaluation layout".
But they seem equivalent and the docs added are great and needed. Thanks.

The only "motivation" is I had to rewrite this part from scratch and this is the first thing I coded. I don't know whether there's a reason to prefer one layout over another. But if you have a reason then perhaps we could change it.

File: vid/src/advz.rs (hunk 1 above)
Comment (ggutoski) [reply vid/src/advz.rs:514]: done in https://github.com/EspressoSystems/jellyfish/pull/678/commits/7a37893155ec163efa6a8e5553a3559609848c70

File: vid/src/advz/test.rs (hunk 2 above)
Comment (ggutoski) [reply vid/src/advz/test.rs:275]: done in https://github.com/EspressoSystems/jellyfish/pull/678/commits/52e7c17e87bf4b7c88ef0703163684dfb054fa14. All tests now use nontrivial multiplicity. (Set in `advz_init`.)

**Review by akonring — APPROVED** [review]

**Pull request merged**